### PR TITLE
simplify filter ASTs in frontend

### PR DESCRIPTION
**Pull request by WebFreak001**

### CODE REVIEW

File: kraken_frontend/src/utils/filter/parser.ts [hunk 1]
```diff
@@ -382,50 +88,56 @@ function parseAst(input: string, parseColumn: (column: string, cursor: Cursor) =
...
 
 /** Parse an {@link Expr.And} expression using a `parseValue` to parse the leaves */
-function parseAnd<T>(tokens: Cursor, parseValue: (cursor: Cursor) => Expr.Value<T>): Expr.And<T> {
-    const list = [parseNot(tokens, parseValue)];
-    while (tokens.peekToken()?.type === "logicalAnd") {
-        tokens.nextToken(); // Consume the `&`
-        list.push(parseNot(tokens, parseValue));
-    }
-    return { and: list };
+export function parseAnd<T>(parseValue: (cursor: Cursor) => Expr.Value<T>): (tokens: Cursor) => Expr.And<T> {
```
Comment (gammelalf) [inline kraken_frontend/src/utils/filter/parser.ts:103]: why does this return a lambda?
Isn't it only ever called in `parseOr` which immediately calls it exactly once?

File: kraken_frontend/src/utils/filter/parser.ts [hunk 2]
```diff
@@ -382,50 +88,56 @@ function parseAst(input: string, parseColumn: (column: string, cursor: Cursor) =
...
 
 /** Parse a {@link Expr.Not} using a `parseValue` to parse the potentially negated value */
-function parseNot<T>(tokens: Cursor, parseValue: (cursor: Cursor) => Expr.Value<T>): Expr.Not<T> {
-    let not = false;
-    if (tokens.peekToken()?.type === "logicalNot") {
-        tokens.nextToken(); // Consume the `!`
-        not = true;
-    }
-    return { not, value: parseValue(tokens) };
+export function parseNot<T>(parseValue: (cursor: Cursor) => Expr.Value<T>): (tokens: Cursor) => Expr.Not<T> {
```
Comment (gammelalf) [inline kraken_frontend/src/utils/filter/parser.ts:115]: why does this return a lambda?
Isn't it only ever called in `parseAnd` which immediately calls it exactly once?

File: kraken_frontend/src/utils/filter/parser.ts [hunk 3]
```diff
@@ -382,50 +88,56 @@ function parseAst(input: string, parseColumn: (column: string, cursor: Cursor) =
...
 
 /** Parse an {@link Expr.Or} expression using a `parseValue` to parse the leaves */
-function parseOr<T>(tokens: Cursor, parseValue: (cursor: Cursor) => Expr.Value<T>): Expr.Or<T> {
-    const list = [parseAnd(tokens, parseValue)];
-    while (tokens.peekToken()?.type === "logicalOr") {
-        tokens.nextToken(); // Consume the `,`
-        list.push(parseAnd(tokens, parseValue));
-    }
-    return { or: list };
+export function parseOr<T>(parseValue: (cursor: Cursor) => Expr.Value<T>): (tokens: Cursor) => Expr.Or<T> {
```
Comment (gammelalf) [inline kraken_frontend/src/utils/filter/parser.ts:91]: why does this return a lambda?
It is used to wrap every `parseValue` in the declaration of the asts.
But there it is unconditionally wrapped around every single field.
So why don't remove the `parseOr` entirely from the ast declarations and only use it as normal function internally?

**Review by gammelalf — COMMENTED** [review]
As a typescript enthusiast I totally approve the base idea.

However a consideration for @myOmikron: The previous design tried to make the rust and typescript code base as similar as possible, because every change has to be made to both and a contributor therefore has to understand both.

File: kraken_frontend/src/utils/filter/parser.ts (hunk 3 above)
Comment (WebFreak001) [reply kraken_frontend/src/utils/filter/parser.ts:91]: this is so the code how it was previously written is still recognizable the same way now: `parseOr(tokens, parseFoo(parseBar(...)))` how it was used in the manual switch-case parsing code before now simply became `parseOr(parseFoo(parseBar(...)))` inside the declaration. I don't think we will ever have anything else than or on top level, but to match the previous code I kept it like that. However if we want to diverge from the original code more we can remove the unused / repeating stuff like this.

Comment (gammelalf) [reply kraken_frontend/src/utils/filter/parser.ts:91]: The old code's structure was to mimic rust's, but this PR abandons that anyway.
So I'd prefer it without the meaningless `parseOr` everywhere.

**Pull request merged**